File: jback-api/src/main/java/ch/patchcode/jback/api/registration/InitialRegistrationData.java
package ch.patchcode.jback.api.registration;

import ch.patchcode.jback.api.verification.VerificationMean;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import org.inferred.freebuilder.FreeBuilder;

/**
 * This is the record expected from people that want to register.
 * They have to provide their name, and some contact media.
 */
@ApiModel("personal information for registration")
@FreeBuilder
public abstract class InitialRegistrationData {

    @ApiModelProperty
    public abstract String getFirstName();

    @ApiModelProperty
    public abstract String getLastName();

    @ApiModelProperty
    public abstract VerificationMean getVerificationMean();

    @JsonCreator
    public static InitialRegistrationData create(
            @JsonProperty("firstName") String firstName,
            @JsonProperty("lastName") String lastName,
            @JsonProperty("verificationMean") VerificationMean verificationMean
    ) {

        return new Builder()
                .setFirstName(firstName)
                .setLastName(lastName)
                .setVerificationMean(verificationMean)
                .build();
    }

    public ch.patchcode.jback.security.secBaseImpl.InitialRegistrationData toDomain() {
        return new ch.patchcode.jback.security.secBaseImpl.InitialRegistrationData.Builder()
                .setFirstName(getFirstName())
                .setLastName(getLastName())
                .setVerificationMean(getVerificationMean().toDomain())
                .build();
    }

    public static class Builder extends InitialRegistrationData_Builder {
    }
}

File: jback-angular/src/main/ngapp/src/app/registration/initial-registration-data.ts
import { VerificationMean } from '../verification-means/verification-mean';

export class InitialRegistrationData {

  firstName: string = null;

  lastName: string = null;

  verificationMean: VerificationMean = null;

  private notNullOrEmpty(str: string): boolean {
    return typeof str === 'string' && str !== '';
  }

  isValid(): boolean {
    return this.notNullOrEmpty(this.firstName) &&
      this.notNullOrEmpty(this.lastName) &&
      this.verificationMean instanceof VerificationMean &&
      this.verificationMean.isValid();
  }
}
